p2p/http/libp2phttp.go

```diff
@@ -0,0 +1,756 @@
+
+// ProtocolMeta is metadata about a protocol.
+type ProtocolMeta struct {
+	// Path defines the HTTP Path prefix used for this protocol
+	Path string `json:"path"`
+}
+
+type PeerMeta map[protocol.ID]ProtocolMeta
+
+// WellKnownHandler is an http.Handler that serves the .well-known/libp2p resource
+type WellKnownHandler struct {
+	wellknownMapMu   sync.Mutex
+	wellKnownMapping PeerMeta
+}
+
+// streamHostListen retuns a net.Listener that listens on libp2p streams for HTTP/1.1 messages.
+func streamHostListen(streamHost host.Host) (net.Listener, error) {
+	return gostream.Listen(streamHost, ProtocolIDForMultistreamSelect)
+}
+
+func (h *WellKnownHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Check if the requests accepts JSON
+	accepts := r.Header.Get("Accept")
+	if accepts != "" && !(strings.Contains(accepts, "application/json") || strings.Contains(accepts, "*/*")) {
+		http.Error(w, "Only application/json is supported", http.StatusNotAcceptable)
+		return
+	}
+
+	if r.Method != http.MethodGet {
+		http.Error(w, "Only GET requests are supported", http.StatusMethodNotAllowed)
+		return
+	}
+
+	// Return a JSON object with the well-known protocols
+	h.wellknownMapMu.Lock()
+	mapping, err := json.Marshal(h.wellKnownMapping)
+	h.wellknownMapMu.Unlock()
+	if err != nil {
+		http.Error(w, "Marshal error", http.StatusInternalServerError)
+		return
+	}
+	w.Header().Add("Content-Type", "application/json")
+	w.Header().Add("Content-Length", strconv.Itoa(len(mapping)))
+	w.Write(mapping)
+}
+
+func (h *WellKnownHandler) AddProtocolMapping(p protocol.ID, protocolMeta ProtocolMeta) {
+	h.wellknownMapMu.Lock()
+	if h.wellKnownMapping == nil {
+		h.wellKnownMapping = make(map[protocol.ID]ProtocolMeta)
+	}
+	h.wellKnownMapping[p] = protocolMeta
+	h.wellknownMapMu.Unlock()
+}
+
+func (h *WellKnownHandler) RemoveProtocolMapping(p protocol.ID) {
+	h.wellknownMapMu.Lock()
+	if h.wellKnownMapping != nil {
+		delete(h.wellKnownMapping, p)
+	}
+	h.wellknownMapMu.Unlock()
+}
+
+// Host is a libp2p host for request/responses with HTTP semantics. This is
+// in contrast to a stream-oriented host like the core host.Host interface. Its
+// zero-value (&Host{}) is usable. Do not copy by value.
+// See examples for usage.
+//
+//	Warning, this is experimental. The API will likely change.
+type Host struct {
+	// StreamHost is a stream based libp2p host used to do HTTP over libp2p streams. May be nil
+	StreamHost host.Host
+	// ListenAddrs are the requested addresses to listen on. Multiaddrs must be
+	// a valid HTTP(s) multiaddr. Only multiaddrs for an HTTP transport are
+	// supported (must end with /http or /https).
+	ListenAddrs []ma.Multiaddr
+	// TLSConfig is the TLS config for the server to use
+	TLSConfig *tls.Config
+	// ServeInsecureHTTP indicates if the server should serve unencrypted HTTP requests over TCP.
+	ServeInsecureHTTP bool
+	// ServeMux is the http.ServeMux used by the server to serve requests
+	ServeMux http.ServeMux
```

Comment:
No it will not use http.DefaultServeMux. To me that's a default for the standard library, and I think it would be confusing if we referenced it here. Users would create a new host that automatically exposes things they weren't expecting.

The reason this is public is to allow a user to set a custom servemux, or to manually manage the mapping outside of `SetHTTPHandler`. I'll expand the comment to mention this.

Reply:
This should probably be a pointer if we want to let users set a custom servemux